`testGetFreePort` fails in WSL on Windows

## Summary
For some reason, on Java 8 in Ubuntu 18.04 running in WSL, the DS 3.x tests are failing for `testGetFreePort`. It looks like the JRE or Windows are allowing Java to bind two different server sockets to the same port (port sharing), which violates the test's setup expectations.

## Steps To Reproduce ("Repro Steps")
1. Checkout `sustaining/3.0`.
2. Build the project with `mvn clean install`.

## Expected Result (Behavior You Expected to See)
The build should succeed with no test failures.

## Actual Result (Behavior You Saw)
The build fails with this error:
<details>
  <summary>Full failure log</summary>
  <pre>
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running TestSuite
Configuring TestNG with: TestNG652Configurator

Test environment:

  Java version: 1.8.0_191
  Java vendor:  Oracle Corporation
  JVM name:     OpenJDK 64-Bit Server VM
  JVM version:  25.191-b12
  JVM vendor:   Oracle Corporation
  JVM info:     mixed mode
  Java home:    /usr/lib/jvm/java-8-openjdk-amd64/jre
  OS:           Linux 4.4.0-17763-Microsoft amd64
  Processors:   4
  Max memory:   2837446656
  Total memory: 192937984

How to read the progressive status info:

  Test duration status: {Total min:sec.  Since last status sec.}
  Test count status:  {# test classes  # test methods  # test method invocations  # test failures}.
  TestClass (the class that just completed)

{ 0:00 (  0s)}  {  0c    0m     0i 0f}  : starting
{ 0:00 (  0s)}  {  1c    6m     6i 0f}  : CertificateTestCase
{ 0:00 (  0s)}  {  2c    9m     9i 0f}  : ProductInformationTest
{ 0:00 (  0s)}  {  3c   13m    21i 0f}  : SetupCliTestCase
{ 0:00 (  0s)}  {  4c   16m    24i 0f}  : DataConfigurationTestCase



                 T E S T   F A I L U R E ! ! !

Failed Test:  org.forgerock.opendj.server.setup.model.ListenerSettingsTestCase#testGetFreePort
Failure Cause:  java.lang.AssertionError: actual value:<61728> should not be equal to:<61728>
    org.fest.assertions.Fail.failure(Fail.java:228)
    org.fest.assertions.Fail.fail(Fail.java:218)
    org.fest.assertions.Fail.failIfEqual(Fail.java:53)
    org.fest.assertions.GenericAssert.isNotEqualTo(GenericAssert.java:228)
    org.fest.assertions.IntAssert.isNotEqualTo(IntAssert.java:71)
    org.forgerock.opendj.server.setup.model.ListenerSettingsTestCase.testGetFreePort(ListenerSettingsTestCase.java:80)


-------------------------------------------------------------------------------


{ 0:00 (  0s)}  {  5c   20m    28i 1f}  : ListenerSettingsTestCase
{ 0:00 (  0s)}  {  6c   38m    46i 1f}  : ModelTestCase
{ 0:00 (  0s)}  {  7c   43m    51i 1f}  : RuntimeOptionsTestCase
The following unit tests failed:
  org.forgerock.opendj.server.setup.model.ListenerSettingsTestCase#testGetFreePort


Include the ant option '-Dtest.failures=true' to rerun only the failed tests.

Wrote full test report to:
/mnt/c/Users/guypa/Documents/WrenProjects/wrends/opendj-server/target/surefire-reports/results.txt
Test classes run interleaved: 0
Final amount of memory in use: 3.0 MB
Final number of threads: 1

Tests run: 51, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.897 sec <<< FAILURE! - in TestSuite
testGetFreePort(org.forgerock.opendj.server.setup.model.ListenerSettingsTestCase)  Time elapsed: 0.06 sec  <<< FAILURE!
java.lang.AssertionError: actual value:<61728> should not be equal to:<61728>
        at org.fest.assertions.Fail.failure(Fail.java:228)
        at org.fest.assertions.Fail.fail(Fail.java:218)
        at org.fest.assertions.Fail.failIfEqual(Fail.java:53)
        at org.fest.assertions.GenericAssert.isNotEqualTo(GenericAssert.java:228)
        at org.fest.assertions.IntAssert.isNotEqualTo(IntAssert.java:71)
        at org.forgerock.opendj.server.setup.model.ListenerSettingsTestCase.testGetFreePort(ListenerSettingsTestCase.java:80)


Results :

Failed tests:
  ListenerSettingsTestCase.testGetFreePort:80 actual value:<61728> should not be equal to:<61728>

Tests run: 51, Failures: 1, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 6.884 s
[INFO] Finished at: 2019-04-18T14:51:48Z
[INFO] Final Memory: 27M/252M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.18.1:test (default-test) on project opendj-server: There are test failures.
[ERROR]
[ERROR] Please refer to /mnt/c/Users/guypa/Documents/WrenProjects/wrends/opendj-server/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
</pre>
</details>

## Environment
- **Product version (include GIT commit ID if using a dev build):** 3.0.0 @ 3a60e66
- **Operating system:**
   ```
   Ubuntu 18.04.1 LTS
   Linux 4.4.0-17763-Microsoft #379-Microsoft Wed Mar 06 19:16:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux
   ```
- **JRE version:**
   ```
   openjdk version "1.8.0_191"
   OpenJDK Runtime Environment (build 1.8.0_191-8u191-b12-2ubuntu0.18.04.1-b12)
   OpenJDK 64-Bit Server VM (build 25.191-b12, mixed mode)
   ```
- **Was instance upgraded from a previous version? If so, what version were you using before?** No

## Additional Notes
Looking into this issue revealed a separate issue with the way that the DS 3.x SDK looks for a free port. A separate issue for that is pending.

--- a/opendj-server/src/test/java/org/forgerock/opendj/server/setup/model/ListenerSettingsTestCase.java
+++ b/opendj-server/src/test/java/org/forgerock/opendj/server/setup/model/ListenerSettingsTestCase.java
@@ -22,6 +22,7 @@
  *
  *
  *      Copyright 2014 ForgeRock AS.
+ *      Portions Copyright 2019 Wren Security.
  */
 package org.forgerock.opendj.server.setup.model;
 
@@ -72,13 +73,23 @@ public class ListenerSettingsTestCase extends AbstractSetupTestCase {
     public void testGetFreePort() throws Exception {
         // Finds a free socket
         final InetSocketAddress isa = TestCaseUtils.findFreeSocketAddress();
-        // Bound the free socket
-        final int port = isa.getPort();
-        final ServerSocket boundSocket = new ServerSocket(port);
-        // Verify the new port number is different from the free socket and verify it's free.
-        final int newPort = ListenerSettings.getFreeSocketPort(port);
-        assertThat(newPort).isNotEqualTo(port);
-        assertTrue(Math.abs(newPort - port) % PORT_INCREMENT == 0);
+
+        // Bind the free socket
+        final int boundPort = isa.getPort();
+        final ServerSocket boundSocket = new ServerSocket();
+
+        // NOTE: This seems to work for JDK 8+; but, in JDK 9+, the next code line should switch to
+        // being the following instead (which is more precise):
+        //
+        // boundSocket.setOption(StandardSocketOptions.SO_REUSEPORT, false);
+        //
+        boundSocket.setReuseAddress(false);
+        boundSocket.bind(new InetSocketAddress(boundPort));
+
+        // Verify the new port number is different from the bound socket and verify it's free.
+        final int newPort = ListenerSettings.getFreeSocketPort(boundPort);
+        assertThat(newPort).isNotEqualTo(boundPort);
+        assertTrue(Math.abs(newPort - boundPort) % PORT_INCREMENT == 0);
 
         boundSocket.close();
     }

